### PR TITLE
Java 21 migration dev wrok for efs-submission-api

### PR DESCRIPTION
**Pull request by mnycz-ch**
https://companieshouse.atlassian.net/browse/JU-57

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -18,46 +18,53 @@
...
+        <log4j-bom.version>3.0.0-alpha1</log4j-bom.version>
+        <structured-logging.version>3.0.8</structured-logging.version>
         <avro.version>1.11.3</avro.version>
-        <ch-kafka.version>1.4.4</ch-kafka.version>
-        <kafka-models.version>1.0.12</kafka-models.version>
-        <ojdbc6.version>11.2.0.3</ojdbc6.version>
+        <ch-kafka.version>3.0.2</ch-kafka.version>
+        <kafka-clients.version>3.7.0</kafka-clients.version>
+        <ojdbc11.version>23.4.0.24.05</ojdbc11.version>
         <junit-jupiter.version>5.7.1</junit-jupiter.version>
```
Comment (hwatkins-ch) [inline pom.xml:36]: Latest version of junit-jupiter is [5.10.2](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api/5.10.2)

File: pom.xml [hunk 2]
```diff
@@ -18,46 +18,53 @@
...
         <!-- Maven -->
-        <maven-surefire-plugin.version>2.22.2</maven-surefire-plugin.version>
-        <maven-failsafe-plugin.version>2.22.2</maven-failsafe-plugin.version>
-        <maven-compiler-plugin.version>3.8.1</maven-compiler-plugin.version>
-        <jacoco-maven-plugin.version>0.8.5</jacoco-maven-plugin.version>
+        <maven-surefire-plugin.version>3.2.5</maven-surefire-plugin.version>
+        <maven-failsafe-plugin.version>3.2.5</maven-failsafe-plugin.version>
+        <maven-compiler-plugin.version>3.13.0</maven-compiler-plugin.version>
+        <jacoco-maven-plugin.version>0.8.12</jacoco-maven-plugin.version>
+        <jib-maven-plugin.version>3.2.1</jib-maven-plugin.version>
```
Comment (hwatkins-ch) [inline pom.xml:61]: Latest version of jib plugin is [3.4.3](https://mvnrepository.com/artifact/com.google.cloud.tools/jib-maven-plugin/3.4.3)

File: pom.xml [hunk 3]
```diff
@@ -241,13 +252,13 @@
             <!-- update from version in transient dependency of private-api-sdk-java -->
             <groupId>com.google.oauth-client</groupId>
             <artifactId>google-oauth-client</artifactId>
-            <version>1.34.1</version>
+            <version>1.36.0</version>
```
Comment (hwatkins-ch) [inline pom.xml:255]: Put this in a property

File: pom.xml [hunk 4]
```diff
@@ -344,15 +355,17 @@
...
             <artifactId>mockserver-client-java</artifactId>
-            <version>5.5.4</version>
+            <version>${mockserver-client-java.version}</version>
             <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>com.amazonaws</groupId>
             <artifactId>aws-java-sdk-core</artifactId>
-            <version>1.12.343</version>
+            <version>1.12.740</version>
```
Comment (hwatkins-ch) [inline pom.xml:364]: Put this in a property

File: pom.xml [hunk 5]
```diff
@@ -432,7 +448,7 @@
             <!-- See Workaround Note above -->
             <plugin>
                 <artifactId>maven-antrun-plugin</artifactId>
-                <version>1.8</version>
+                <version>3.1.0</version>
```
Comment (hwatkins-ch) [inline pom.xml:451]: Put this in a property

File: pom.xml [hunk 6]
```diff
@@ -18,46 +18,53 @@
...
+        <structured-logging.version>3.0.8</structured-logging.version>
         <avro.version>1.11.3</avro.version>
-        <ch-kafka.version>1.4.4</ch-kafka.version>
-        <kafka-models.version>1.0.12</kafka-models.version>
-        <ojdbc6.version>11.2.0.3</ojdbc6.version>
+        <ch-kafka.version>3.0.2</ch-kafka.version>
+        <kafka-clients.version>3.7.0</kafka-clients.version>
+        <ojdbc11.version>23.4.0.24.05</ojdbc11.version>
         <junit-jupiter.version>5.7.1</junit-jupiter.version>
         <mockito-junit-jupiter.version>3.8.0</mockito-junit-jupiter.version>
```
Comment (hwatkins-ch) [inline pom.xml:37]: Latest version of mockito junit jupiter is [5.12.0](https://mvnrepository.com/artifact/org.mockito/mockito-junit-jupiter/5.12.0)

File: pom.xml [hunk 7]
```diff
@@ -165,12 +177,12 @@
...
             <groupId>org.apache.commons</groupId>
             <artifactId>commons-compress</artifactId>
-            <version>1.24.0</version>
+            <version>${commons-compress.version}</version>
         </dependency>
         <dependency>
             <groupId>org.json</groupId>
             <artifactId>json</artifactId>
-            <version>20231013</version>
+            <version>20240303</version>
```
Comment (hwatkins-ch) [inline pom.xml:185]: Put this in a property

File: pom.xml [hunk 8]
```diff
@@ -181,30 +193,29 @@
...
+                    <artifactId>plexus-sec-dispatcher</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.sonarsource.scanner.api</groupId>
+                    <artifactId>sonar-scanner-api</artifactId>
+                </exclusion>
+            </exclusions>
         </dependency>
         <dependency>
             <groupId>software.amazon.awssdk</groupId>
```
Comment (hwatkins-ch) [inline pom.xml:221]: Can we add a version for this to make sure its latest?

File: src/test/java/uk/gov/companieshouse/efs/api/events/service/fesloader/BatchDaoTest.java [hunk 9]
```diff
@@ -49,7 +49,9 @@ void testBatchDaoObtainsNextBatchIdFromSequence() {
     @Test
     void testBatchDaoObtainsNextBatchNameIdFromSequence() {
         //given
-        when(this.template.queryForObject(anyString(), eq(Long.class), any())).thenReturn(BATCH_NAME_ID);
+        when(this.template.queryForObject(anyString(), eq(Long.class), any(Object[].class))).thenReturn(BATCH_NAME_ID);
+
+        //doReturn(BATCH_NAME_ID).when(this.template.queryForObject(anyString(), eq(Long.class), any()));
```
Comment (hwatkins-ch) [inline src/test/java/uk/gov/companieshouse/efs/api/events/service/fesloader/BatchDaoTest.java:54]: Any reason this is commented out

**Review by hwatkins-ch — CHANGES_REQUESTED** [review]
Can you double check your POM and make sure everything is in property and that we are on the latest versions of dependencies, where we can be. I commented on a few places where I noticed we are not on the latest, but I haven't checked them all. Please double check.

Looks good other than that.

File: src/test/java/uk/gov/companieshouse/efs/api/events/service/fesloader/BatchDaoTest.java (hunk 9 above)
Comment (mnycz-ch) [reply src/test/java/uk/gov/companieshouse/efs/api/events/service/fesloader/BatchDaoTest.java:54]: I was try to replace not working when() with doReturn() and forget to remove after that

File: pom.xml (hunk 8 above)
Comment (mnycz-ch) [reply pom.xml:221]: can not add version to exclusion

File: pom.xml (hunk 7 above)
Comment (mnycz-ch) [reply pom.xml:185]: done

File: pom.xml (hunk 6 above)
Comment (mnycz-ch) [reply pom.xml:37]: mockito-inline latest version is 5.2.0 - updated

File: pom.xml (hunk 5 above)
Comment (mnycz-ch) [reply pom.xml:451]: done

File: pom.xml (hunk 4 above)
Comment (mnycz-ch) [reply pom.xml:364]: done

File: pom.xml (hunk 3 above)
Comment (mnycz-ch) [reply pom.xml:255]: done

File: pom.xml (hunk 2 above)
Comment (mnycz-ch) [reply pom.xml:61]: done

File: pom.xml (hunk 1 above)
Comment (mnycz-ch) [reply pom.xml:36]: done

**Review by hwatkins-ch — APPROVED** [review]

File: src/it/java/uk/gov/companieshouse/efs/api/payment/controller/PaymentControllerDatabaseIT.java [hunk 10]
```diff
@@ -2,14 +2,18 @@
 
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.contains;
-import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.*;
```
Comment (foman-ch) [inline src/it/java/uk/gov/companieshouse/efs/api/payment/controller/PaymentControllerDatabaseIT.java:5]: We try to avoid using wildcard imports

**Review by foman-ch — CHANGES_REQUESTED** [review]

File: src/it/java/uk/gov/companieshouse/efs/api/payment/controller/PaymentControllerDatabaseIT.java (hunk 10 above)
Comment (mnycz-ch) [reply src/it/java/uk/gov/companieshouse/efs/api/payment/controller/PaymentControllerDatabaseIT.java]: done

**Review by foman-ch — APPROVED** [review]

**Pull request merged**